delphi-frontend/src/main/java/au/com/integradev/delphi/symbol/resolve/NameResolutionHelper.java

```diff
@@ -517,18 +523,18 @@ private static boolean handleRoutineReference(
 
     if (parent instanceof AssignmentStatementNode) {
       ExpressionNode assignee = ((AssignmentStatementNode) parent).getAssignee();
-      if (expression == assignee) {
-        return false;
-      }
-
-      if (assignee.getType().isProcedural()) {
-        NameResolver clone = new NameResolver(resolver);
-        clone.disambiguateRoutineReference((ProceduralType) assignee.getType());
-        if (!clone.getDeclarations().isEmpty()) {
-          clone.addToSymbolTable();
-          return true;
-        }
-      }
+      return expression != assignee && handleRoutineReference(assignee.getType(), resolver);
+    } else if (parent instanceof VarDeclarationNode) {
+      return handleRoutineReference(((VarDeclarationNode) parent).getType(), resolver);
+    } else if (parent instanceof ConstDeclarationNode) {
+      return handleRoutineReference(((ConstDeclarationNode) parent).getTypeNode(), resolver);
+    } else if (parent instanceof VarStatementNode) {
+      return handleRoutineReference(((VarStatementNode) parent).getTypeNode(), resolver);
+    } else if (parent instanceof ConstStatementNode) {
+      return handleRoutineReference(((ConstStatementNode) parent).getTypeNode(), resolver);
```

Comment:
The type node is potentially null for `ConstDeclarationNode`, `VarStatementNode`, and `ConstStatementNode`. This just looks like it will fail to resolve any cases where the type is inferred?

Reply:
That's because it will - you can't assign routine references unless the procedural type is explicitly declared.
When this method returns false, the name resolver falls through to the standard primary expression resolution and tries to continue resolving as if its an invocation with an implicit empty argument list.

Reply:
> you can't assign routine references unless the procedural type is explicitly declared.

Can we add a comment here to that effect? I don't think it's obvious in this code that that is intentional.

Reply:
Sure.